### DISCUSSION TITLE
Dividing function for NanoEls H4

### DISCUSSION THREAD

**Discussion by fmw626** [kachurovskiy/nanoels#119] [Ideas]
Hello everyone,

I wanted to share an exciting idea for the NanoEls project, which I showcased in my [show and tell](https://github.com/kachurovskiy/nanoels/discussions/118). I have built a dividing attachment on the back of my lathe spindle. The dividing attachment incorporates a relatively large 3D-printed worm wheel and a worm positioned on a shaft. It can be easily engaged or disengaged with a single nut, allowing for manual operation at the moment. However, my ultimate goal is to integrate this dividing function into the NanoEls H4.

Why the dividing function? Well, it offers the possibility to add rotary features to turned parts without removing them from the chuck, thus preserving concentricity. By attaching a live tool (I currently use a Proxxon rotary tool) to the cross-slide, I can add milled features to turned parts. This feature is incredibly versatile, making it ideal for tasks such as gear making, adding screw holes, creating slots, and much more. 

Now, let's discuss how we can implement this functionality with the NanoEls H4:

Hardware: To enable the dividing function, we would need to add a stepper motor to the worm shaft and connect it to the H4 controller. This stepper motor would thus control the rotation of the dividing attachment.

Software: Since this axis is rotational and not linear, the software controls need to be adjusted accordingly. Here are some ideas for software features:

- Set current angle: This feature allows us to define the starting angle of the dividing attachment (e.g., current angle = 0).
- Go to angle X: With this function, we can command the dividing attachment to rotate to a specific angle (e.g., go to angle 25°).
- Automatic pattern generation: This feature generates a predefined pattern based on a specified angle (e.g., 30° pattern = 0° -> 30° -> 60° -> 90°...). It simplifies the process of creating repetitive rotary features.
- Choice of angle resetting: We can choose whether the angle resets after a full rotation (e.g., 361° becomes 1°) or keeps counting continuously.
- Switch between degrees and fractions: This feature allows us to work with either degrees or fractions (e.g., 90° can be represented as 1/4 rotation).
- Set speed: We can define the speed at which the dividing attachment moves, allowing for quick positioning or slower speeds for precise milling operations.
- Choice of rotation direction: We can select whether the dividing attachment always rotates in the same direction or takes the shortest path to a given position to eliminate backlash.
- Simultaneous control of X and Y axes: By synchronizing the dividing functionality with the X and Y axes, we open up numerous possibilities for CNC machining, such as automatic gear milling.

Integrating these software features into the NanoEls H4 would unlock tremendous potential for enhanced functionality and versatility. What do you think?

![IMG_7182](https://github.com/kachurovskiy/nanoels/assets/122809499/b956c77d-283c-4381-8be6-58e5f5852c8c)
![IMG_7186](https://github.com/kachurovskiy/nanoels/assets/122809499/f0eaf633-0b00-4555-b01c-ed3fa8b41837) ![IMG_7185](https://github.com/kachurovskiy/nanoels/assets/122809499/6445c610-5198-41e5-b54c-03f4ffcf31cd)
![IMG_7174](https://github.com/kachurovskiy/nanoels/assets/122809499/bd0d6f77-4a27-4de2-93da-926b2e044f48)

**Comment by kachurovskiy** [comment]
Hi @fmw626, cool idea, I think we can make it work.

Implementing all the functions you mentioned by itself is not hard, what's hard is making them work intuitively (knowing which button to press without a manual) and mesh well with all the existing functionality (being able to control Z and X and the rotational axis).

Should we call the new axis `C` as this lines up with convention outlined in http://linuxcnc.org/docs/html/gcode/machining-center.html#_axes ?

![buttons](https://github.com/kachurovskiy/nanoels/blob/main/h4/graphics/h4-go-1.png?raw=true)

Looking at the buttons we have, I'm not sure how to make it work best... All buttons are already used for something :) We can add a new mode `HEAD` and repurpose some buttons to new functions when in it. Perhaps we can add more modes to satisfy all the requirements.

Ultimately it could be easier to work with the new axis via the GCode. There it'll be easy to add support for another axis and you can move it alone or with the other 2, set speed, etc. We can also make a fancy HTML UI to control it with as many buttons as we wish :)

Perhaps we should pick 1-2 most important practical tasks involving the new axis and see what's needed for them.

**Reply by fmw626** [reply]
Hi Maxim, i'm glad you find my idea interesting!

> Should we call the new axis C as this lines up with convention outlined in http://linuxcnc.org/docs/html/gcode/machining-center.html#_axes ?

Sounds good!

> Looking at the buttons we have, I'm not sure how to make it work best... All buttons are already used for something :) We can add a new mode HEAD and repurpose some buttons to new functions when in it. Perhaps we can add more modes to satisfy all the requirements.

I completely agree with you regarding the issue of repurposing buttons, as I hadn't considered that aspect in depth yet. Since I don't have hands-on experience with the H4 yet, everything is still hypothetical to me.

> Ultimately it could be easier to work with the new axis via the GCode. There it'll be easy to add support for another axis and you can move it alone or with the other 2, set speed, etc. We can also make a fancy HTML UI to control it with as many buttons as we wish :)

Having an HTML UI with customizable functions is a fantastic concept. It would allow the UI to be modular, enabling each user to select and incorporate the controls they require.

> Perhaps we should pick 1-2 most important practical tasks involving the new axis and see what's needed for them.

In my opinion, the essential functions that should definitely be implemented into the H4 UI itself are as follows:
- Set the current C-axis position as 0°
- Go to a specified number of degrees or enable jogging
- Adjust rotation speed

The other functions are not strictly necessary, but would make the dividing head much more usable. Some settings could be configured before flashing the H4, such as angle resetting and rotation direction behaviors, as you suggested. Alternatively, leaving these settings to be adjusted through the HTML UI should be a good way to go about it.

What are your thoughts on this?
